util/chunk/row_container.go

```diff
@@ -239,17 +282,181 @@ func (a *SpillDiskAction) SetFallback(fallback memory.ActionOnExceed) {
 // SetLogHook sets the hook, it does nothing just to form the memory.ActionOnExceed interface.
 func (a *SpillDiskAction) SetLogHook(hook func(uint64)) {}
 
-// ResetOnce resets the spill action so that it can be triggered next time.
-func (a *SpillDiskAction) ResetOnce() {
+// ResetRowContainer resets the spill action and sets the RowContainer for the SpillDiskAction.
+func (a *SpillDiskAction) ResetRowContainer(c *RowContainer) {
 	a.m.Lock()
 	defer a.m.Unlock()
-	a.once = sync.Once{}
+	a.c = c
+}
+
+// ErrInsertToPartitionFailed indicate that the SortedRowContainer is sorted and prohibit inserting data.
+var ErrInsertToPartitionFailed = terror.ClassExecutor.New(1, "Insert the records unsuccessfully. The SortedRowContainer is sorted.")
+
+// SortedRowContainer provides a place for many rows, so many that we might want to sort and spill them into disk.
+type SortedRowContainer struct {
+	*RowContainer
+	ptrM struct {
+		// RWMutex guarantees the operator for rowPtrs is mutually exclusive.
+		sync.RWMutex
+		// rowPointer store the chunk index and row index for each row.
+		rowPtrs []RowPtr
+	}
+
+	ByItemsDesc []bool
+	// keyColumns is the column index of the by items.
+	keyColumns []int
+	// keyCmpFuncs is used to compare each ByItem.
+	keyCmpFuncs []CompareFunc
+
+	actionSpill *SortAndSpillDiskAction
+}
+
+// NewSortedRowContainer creates a new SortedRowContainer in memory.
+func NewSortedRowContainer(fieldType []*types.FieldType, chunkSize int, ByItemsDesc []bool,
+	keyColumns []int, keyCmpFuncs []CompareFunc) *SortedRowContainer {
+	return &SortedRowContainer{RowContainer: NewRowContainer(fieldType, chunkSize),
+		ByItemsDesc: ByItemsDesc, keyColumns: keyColumns, keyCmpFuncs: keyCmpFuncs}
+}
+
+// Close close the SortedRowContainer
+func (c *SortedRowContainer) Close() error {
+	c.GetMemTracker().Consume(int64(-8 * cap(c.ptrM.rowPtrs)))
+	return c.RowContainer.Close()
+}
+
+func (c *SortedRowContainer) lessRow(rowI, rowJ Row) bool {
+	for i, colIdx := range c.keyColumns {
+		cmpFunc := c.keyCmpFuncs[i]
+		cmp := cmpFunc(rowI, colIdx, rowJ, colIdx)
+		if c.ByItemsDesc[i] {
+			cmp = -cmp
+		}
+		if cmp < 0 {
+			return true
+		} else if cmp > 0 {
+			return false
+		}
+	}
+	return false
+}
+
+// keyColumnsLess is the less function for key columns.
+func (c *SortedRowContainer) keyColumnsLess(i, j int) bool {
+	rowI := c.m.records.GetRow(c.ptrM.rowPtrs[i])
+	rowJ := c.m.records.GetRow(c.ptrM.rowPtrs[j])
+	return c.lessRow(rowI, rowJ)
 }
 
-// ResetOnceAndSetRowContainer resets the spill action and sets the RowContainer for the SpillDiskAction.
-func (a *SpillDiskAction) ResetOnceAndSetRowContainer(c *RowContainer) {
+// InitPointersAndSort inits pointers and sorts the records.
+func (c *SortedRowContainer) InitPointersAndSort() {
```

Comment:
`InitPointers` may confuse the code readers because the caller does not care the unexported `rowPtrs` field.

```suggestion
func (c *SortedRowContainer) Sort() {
```